### PR TITLE
[MRG+2] R diffinv parity

### LINKED ISSUE #180: Any inverse function for pmdarima.utils.diff?
I am just curious that is there any inverse function like `diffinv`  in `R` for pmdarima.utils.diff?

### PR DESCRIPTION
**Pull request by charlesdrotar**
# Description

- **Please include a summary of the change and which issue is fixed.**
  - Adds a `diff_inv` function that is in parity with R.

- **Please also include relevant motivation and context. List any dependencies
that are required for this change.**
   - No dependency changes.

- Fixes #180

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation change

*FYI: There will have to be a corresponding documentation change that shows how diff_inv functions also in relation to the `diff_inv(diff(x))` scenario to clear up that it doesn't just return the original matrix `x` unless `xi` is appropriately specified. See R example [here](https://rdrr.io/r/stats/diffinv.html). If we remove `xi` it will essentially have an offset of the initial value `xi`*

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.
Provide instructions so we can reproduce. Please also list any relevant details
for your test configuration

- Test that `diff_inv(x)` matches R for a 2-d matrix:
  - [x] `lag=1`
  - [x] `lag>1` 
  - [x] `differences=1`
  - [x] `differences>1`

*It should follow that if it works for 2-d the math would work for a 1-d, so we test the more complex case*

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: pmdarima/utils/_array.pyx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    References
+    ----------
+    .. [1] https://github.com/wch/r-source/blob/trunk/src/library/stats/R/diffinv.R#L39
+    .. [2] https://github.com/mirror/r/blob/65a0e33a4b0a119703586fcd1f9742654738ae54/src/library/stats/src/PPsum.c#L46
+    """
+    cdef INTP i, n = x.shape[0]
+    cdef np.ndarray[double, ndim=1, mode='c'] ans = \
+        np.zeros(n + lag, dtype=np.float64, order='c')
+
+    for i in range(lag, lag + n):
```
Comment (tgsmith61591) [inline pmdarima/utils/_array.pyx]: I bet you could get a performance boost by wrapping this in `with nogil:`

File: pmdarima/utils/array.py [hunk 2]
```diff
@@ -313,6 +315,74 @@ def diff(x, lag=1, differences=1):
...
+        for i in range(m):
+            y[:, i] = _diff_inv_vector(x[:, i], lag, differences, xi[:, i])
+
+    return y
+
+
+def diff_inv(x, lag=1, differences=1, xi=None):
+    """
+    Inverse the difference of an array.
+    """
```
Comment (tgsmith61591) [inline pmdarima/utils/array.py]: Could you fill in the docstring to look like the other ones? Ideally including a couple links to R source and some examples. Could you also update `whats_new.rst` and add this to the API reference in doc?

File: pmdarima/utils/array.py [hunk 3]
```diff
@@ -313,6 +315,74 @@ def diff(x, lag=1, differences=1):
...
+    x = check_array(
+        x, dtype=DTYPE, copy=False, force_all_finite=False, ensure_2d=False)
+
+    # R code: if (lag < 1L || differences < 1L)
+    # R code: stop ("bad value for 'lag' or 'differences'")
+    if any(v < 1 for v in (lag, differences)):
+        raise ValueError('lag and differences must be positive (> 0) integers')
```
Comment (tgsmith61591) [inline pmdarima/utils/array.py]: Would you mind a couple tests to ensure we hit the expected Value Errors?

Comment (charlesdrotar) [reply pmdarima/utils/array.py]: Done!

File: pmdarima/utils/array.py (hunk 2 above)
Comment (charlesdrotar) [reply pmdarima/utils/array.py]: Finished the docstring, but forgot to update the `.rst`. Working on that now.

Comment (charlesdrotar) [reply pmdarima/utils/array.py]: Done!

File: pmdarima/utils/_array.pyx (hunk 1 above)
Comment (charlesdrotar) [reply pmdarima/utils/_array.pyx]: Done!

**Pull request merged**